util/fibers/listener_interface.cc

```diff
@@ -170,16 +170,18 @@ void ListenerInterface::RunAcceptLoop() {
   pool_->AwaitFiberOnAll([&](auto* pb) {
     auto it = conn_list.find(this);
     DCHECK(it != conn_list.end());
-    auto* clist = it->second;
+    TLConnList* clist = it->second;
 
-    cur_conn_cnt.fetch_add(clist->list.size(), memory_order_relaxed);
+    // Important to pop from the end because Shutdown may suspend and the list may change
+    while (!clist->list.empty()) {
+      Connection& conn = clist->list.back();
+      clist->list.pop_back();
 
-    // This is correctly atomic because Shutdown() does not block the fiber or yields.
-    // However once we implement Shutdown via io_uring, this code becomes unsafe.
-    for (auto& conn : clist->list) {
+      boost::intrusive_ptr guard(&conn);
       conn.Shutdown();
       DVSOCK(1, conn) << "Shutdown";
     }
+    cur_conn_cnt.fetch_add(clist->list.size(), memory_order_relaxed);
```

Comment:
Isn't `clist->list` empty at this point?

Reply:
no necessarily. Shutdown does not close and removes connections. it just shuts them down and they still need to go through their closing sequence before they are removed from the list

Reply:
But you call `pop_back()` in line 178?

Reply:
oh, it's a bug and I am not sure how to fix it actually.

Reply:
ok, provided an awkward fix.

Reply:
it's an intrusive list, meaning that if an element is not linked then the list was reduced by one element.
Also, `Shutdown` by itself does not cause the element to be removed from the list. Shutdown triggers the finalization flow for the connection, that at the end will remove the connection from the list (at the end of RunSingleConnection function).
But if Shutdown preempts we may or may now have the connection to be unlinked depending how quickly RunSingleConnection finishes its run.

Reply:
the list will never have new elements so any element that is unlinked reduces the list by one, so no endless loop here

Reply:
Where does it remove from the list? I see that in the original code we used to call `pop_back()` here, but in the latest iteration we don't so how does that conform with:

> it's an intrusive list, meaning that if an element is not linked then the list was reduced by one element.

?

Please excuse my picking, I'm just not sure what I'm missing :)

Reply:
It's inside RunSingleConnection.
here: https://github.com/romange/helio/pull/246/files#diff-01d997cdbc4d2835ea398fa879f935f6dee87da00e4fdf7ae7264bdd5a98f7f8R255-R257

Reply:
please note that the goal of the loop is not to empty the list but trigger closing flows, so an element may or may not be removed.